packages/framework/ai-collab/src/explicit-strategy/agentEditReducer.ts

```diff
@@ -0,0 +1,570 @@
+			if (treeSchema.kind === FieldKind.Optional && treeEdit.content === undefined) {
+				if (isRootNode) {
+					treeView.root = treeEdit.content;
+				} else {
+					// eslint-disable-next-line no-param-reassign, @typescript-eslint/no-explicit-any
+					(treeNode as any) = treeEdit.content;
+				}
+			} else {
+				for (const allowedType of treeSchema.allowedTypeSet.values()) {
+					if (schemaIdentifier === allowedType.identifier) {
+						if (isConstructable(allowedType)) {
+							const simpleNodeSchema = allowedType as unknown as new (
+								dummy: unknown,
+							) => TreeNode;
+							const rootNode = new simpleNodeSchema(treeEdit.content);
+							validator?.(rootNode);
+							if (isRootNode) {
+								// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
+								(treeView as any).root = rootNode;
+							} else {
+								// eslint-disable-next-line no-param-reassign
+								treeNode = rootNode;
+							}
+							insertedObject = rootNode;
+						} else {
+							if (isRootNode) {
+								// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
+								(treeView as any).root = treeEdit.content;
+							} else {
+								// eslint-disable-next-line no-param-reassign, @typescript-eslint/no-explicit-any
+								(treeNode as any) = treeEdit.content;
+							}
+						}
+					}
+				}
+			}
+
+			return insertedObject === undefined
+				? treeEdit
+				: {
+						...treeEdit,
+						content: contentWithIds(insertedObject, idGenerator),
+					};
+		}
+		case "insert": {
+			const { array, index } = getPlaceInfo(treeEdit.destination, idGenerator);
+
+			const parentNodeSchema = Tree.schema(array);
+			populateDefaults(treeEdit.content, definitionMap);
+
+			const schemaIdentifier = getSchemaIdentifier(treeEdit.content);
+
+			// We assume that the parentNode for inserts edits are guaranteed to be an arrayNode.
+			const allowedTypes = [
+				...normalizeAllowedTypes(parentNodeSchema.info as ImplicitAllowedTypes),
+			];
+
+			for (const allowedType of allowedTypes.values()) {
+				if (allowedType.identifier === schemaIdentifier && typeof allowedType === "function") {
+					const simpleNodeSchema = allowedType as unknown as new (dummy: unknown) => TreeNode;
+					const insertNode = new simpleNodeSchema(treeEdit.content);
+					validator?.(insertNode);
+					array.insertAt(index, insertNode);
+					return {
+						...treeEdit,
+						content: contentWithIds(insertNode, idGenerator),
+					};
+				}
+			}
+			fail("inserted node must be of an allowed type");
+		}
+		case "remove": {
+			const source = treeEdit.source;
+			if (isObjectTarget(source)) {
+				const node = getNodeFromTarget(source, idGenerator);
+				const parentNode = Tree.parent(node);
+				// Case for deleting rootNode
+				if (parentNode === undefined) {
+					const treeSchema = isRootNode ? treeView.schema : Tree.schema(treeNode);
```

Comment:
It's not possible for isRootNode to be false here, right? If it can be false, then we should be getting Tree.schema(node), right?

Reply:
It's was possible with the addition of the node "scoping", where the ai could use setRoot to replace a node essentially  feature. But, it was half baked and didn't really make sense. The edit description wasn't even updated.

FYI  the 'setRoot' has been removed in commit fb37d5500e19de5dd5a5664513592c997115e358, to be revisited later